server/src/main/java/org/elasticsearch/health/GetHealthAction.java

```diff
@@ -0,0 +1,155 @@
+        }
+
+        public Response(StreamInput in) {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.startObject();
+            builder.field("status", "green");
+            builder.field("cluster_name", clusterName.value());
+            builder.array("impacts");
+            builder.startObject("components");
+            for (Component component : components) {
+                builder.field(component.getName());
+                component.toXContent(builder, params);
+            }
+            builder.endObject();
+            return builder.endObject();
+        }
+    }
+
+    public static class Request extends ActionRequest {
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    public static class TransportAction extends org.elasticsearch.action.support.TransportAction<Request, Response> {
+
+        private final ClusterService clusterService;
+
+        @Inject
+        public TransportAction(
+            final ActionFilters actionFilters,
+            final TransportService transportService,
+            final ClusterService clusterService
+        ) {
+            super(NAME, actionFilters, transportService.getTaskManager());
+            this.clusterService = clusterService;
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+            final ClusterState clusterState = clusterService.state();
+            final Controller component = new Controller(clusterService.localNode(), clusterState);
+            listener.onResponse(new Response(clusterService.getClusterName(), Arrays.asList(component, new Backups())));
+        }
+    }
+
+    public abstract static class Component implements ToXContentObject {
+
+        public abstract String getName();
+
+        public abstract ClusterHealthStatus getStatus();
+
+        public abstract List<Indicator> getIndicators();
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("status", getStatus());
+            builder.startArray("indicators");
+            List<Indicator> indicators = getIndicators();
+            for (Indicator indicator : indicators) {
+                indicator.toXContent(builder, params);
+            }
+            builder.endArray();
+            return builder.endObject();
+        }
+
+    }
+
+    public abstract static class Indicator implements ToXContentObject {
```

Comment:
Same suggestion here. WDYT about making `Indicator` an interface?
```java
public interface Indicator extends ToXContentObject {

    String getName();

    ClusterHealthStatus getStatus();

    String getExplain();

    void writeMeta(XContentBuilder builder, Params params) throws IOException;

    @Override
    default XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
        builder.startObject();
        builder.field("name", getName());
        builder.field("status", getStatus());
        builder.field("explain", getExplain());
        builder.object("meta", xContentBuilder -> writeMeta(builder, params));
        // TODO: Add detail / documentation
        return builder.endObject();
    }
}
```

Reply:
I do not think there is a value in making this abstract class or interface. I believe all indicators could be final classes. Services (or most) services could declare them as compile time constants. Otherwise we might end up with lots of very similar pojo like implementations

Reply:
I mean we need variable behavior for the writing metadata. We could move the abstraction to some type of consumer passed in or something. But I find the abstract class to be cleaner? I don't really see the current objection to different classes for indicators. It allows easy find implementations in code. And allows certain patterns to be adopted like NAME, EXPLAIN constants. Along with separate indicators out to dedicated packages by components (the current `controller` package).

Reply:
I would find it nicer to let each indicator have a `Service` that can calculate it's status.

`Indicator` would be pure data, could be a final class with no subclasses. It could have a `reason` field containing the reason (which would be an interface or abstract class).

I think the Service should go into a separate PR. But the reason vs indicator could go here.

Reply:
This idea is reflected in https://github.com/elastic/elasticsearch/pull/83205

Reply:
I made this change. No abstraction, but that can be made when we add more indicators.